File: src/test/resources/de/hhu/stups/codegenerator/refinement/BInteger.cpp
#include <iostream>
#include <string>
#include "BBoolean.cpp"

#ifndef BINTEGER_H
#define BINTEGER_H

using namespace std;

class BInteger : public BObject {
	private:
	    int value;

    public:

        typedef BInteger current_type;
        typedef void value_type;
        typedef void left_type;
        typedef void right_type;

        BInteger(const int& val) {
            value = val;
        }

        int intValue() const {
            return value;
        }

        BInteger(){};

        BInteger(const BInteger& val) {
            value = val.value;
        }

        int compareTo(const BInteger& o) const {
            return value - o.value;
        }

        BBoolean lessEqual(const BInteger& o) const {
            return compareTo(o) <= 0;
        }


        BBoolean greaterEqual(const BInteger& o) const {
            return compareTo(o) >= 0;
        }

        BBoolean less(const BInteger& o) const {
            return compareTo(o) < 0;
        }

        BBoolean greater(const BInteger& o) const {
            return compareTo(o) > 0;
        }

        BBoolean equal(const BInteger& o) const {
            return compareTo(o) == 0;
        }

        BBoolean unequal(const BInteger& o) const {
            return compareTo(o) != 0;
        }

        BInteger plus(const BInteger& o) const {
            return value + o.value;
        }

        BInteger minus(const BInteger& o) const {
            return value - o.value;
        }

        BInteger multiply(const BInteger& o) const {
            return value * o.value;
        }

        BInteger power(const BInteger& o) const {
            return value ^ o.value;
        }

        BInteger divide(const BInteger& o) const {
            return value / o.value;
        }

        BInteger modulo(const BInteger& o) const {
            return value % o.value;
        }
        BInteger succ() const {
            return value + 1;
        }

        BInteger pred() const {
            return value - 1;
        }

        BInteger negative() const {
            return -value;
        }

        BInteger positive() const {
            return *this;
        }

        friend bool operator !=(const BInteger& o1, const BInteger& o2) {
            return o1.value != o2.value;
        }

        friend bool operator ==(const BInteger& o1, const BInteger& o2) {
            return o1.value == o2.value;
        }

        void operator =(const BInteger& other) {
            this->value = other.value;
        }

        int hashCode() const {
            int prime = 31;
            int result = 1;
            result = prime * result + value;
            return result;
        }

        friend std::ostream& operator<<(std::ostream &strm, const BInteger &b) {
          return strm << b.value;
        }

        BBoolean isInteger() {
            return BBoolean(true);
        }

        BBoolean isNotInteger() {
            return BBoolean(false);
        }

        BBoolean isNatural() {
            return this->greaterEqual(BInteger(0));
        }

        BBoolean isNotNatural() {
            return this->isNatural()._not();
        }

        BBoolean isNatural1() {
            return this->greater(BInteger(0));
        }

        BBoolean isNotNatural1() {
            return this->isNatural1()._not();
        }
};
#endif

File: btypes_primitives/src/main/python_magicstack_immutable/btypes/BInteger.py
from btypes.BBoolean import *


class BInteger:

    def __init__(self, value):
        self.__value = value

    def __add__(self, other):
        if type(other) == str:
            return str(self) + other
        return BInteger(self.__value + other.__value)

    def __radd__(self, other):
        if type(other) == str:
            return other + str(self)
        return BInteger(self.__value + other.__value)

    def __sub__(self, other: 'BInteger') -> 'BInteger':
        return BInteger(self.__value - other.__value)

    def __mul__(self, other: 'BInteger') -> 'BInteger':
        return BInteger(self.__value * other.__value)

    def __mod__(self, other: 'BInteger') -> 'BInteger':
        return BInteger(self.__value % other.__value)

    def __div__(self, other: 'BInteger') -> 'BInteger':
        return BInteger(self.__value // other.__value)

    def __neg__(self) -> 'BInteger':
        return BInteger(-self.__value)

    def __lt__(self, other: 'BInteger') -> 'bool':
        return self.__value < other.__value

    def __le__(self, other: 'BInteger') -> 'bool':
        return self.__value <= other.__value

    def __eq__(self, other: 'BInteger') -> 'bool':
        if not isinstance(other, BInteger):
            return False
        return self.__value == other.__value

    def __ne__(self, other: 'BInteger') -> 'bool':
        return self.__value != other.__value

    def __gt__(self, other: 'BInteger') -> 'bool':
        return self.__value > other.__value

    def __ge__(self, other: 'BInteger') -> 'bool':
        return self.__value >= other.__value

    def __pow__(self, other: 'BInteger') -> 'BInteger':
        return self.__value ** other.__value

    def __str__(self) -> 'str':
        return str(self.__value)

    def plus(self, other: 'BInteger') -> 'BInteger':
        return self.__add__(other)

    def minus(self, other: 'BInteger') -> 'BInteger':
        return self.__sub__(other)

    def multiply(self, other: 'BInteger') -> 'BInteger':
        return self.__mul__(other)

    def modulo(self, other: 'BInteger') -> 'BInteger':
        return self.__mod__(other)

    def divide(self, other: 'BInteger') -> 'BInteger':
        return self.__div__(other)

    def negative(self) -> 'BInteger':
        return self.__neg__()

    def less(self, other: 'BInteger') -> 'BBoolean':
        return BBoolean(self.__value < other.__value)

    def lessEqual(self, other: 'BInteger') -> 'BBoolean':
        return BBoolean(self.__value <= other.__value)

    def equal(self, other: 'BInteger') -> 'BBoolean':
        return BBoolean(self.__value == other.__value)

    def unequal(self, other: 'BInteger') -> 'BBoolean':
        return BBoolean(self.__value != other.__value)

    def greater(self, other: 'BInteger') -> 'BBoolean':
        return BBoolean(self.__value > other.__value)

    def greaterEqual(self, other: 'BInteger') -> 'BBoolean':
        return BBoolean(self.__value >= other.__value)

    def succ(self) -> 'BInteger':
        return BInteger(self.__value + 1)

    def pred(self) -> 'BInteger':
        return BInteger(self.__value - 1)

    def power(self, other: 'BInteger') -> 'BInteger':
        return self.__pow__(other)

    def positive(self):
        return self

    def intValue(self):
        return self.__value

    def isInteger(self) -> 'BBoolean':
        return BBoolean(true)

    def isNotInteger(self) -> 'BBoolean':
        return BBoolean(false)

    def isNatural(self) -> 'BBoolean':
        return self.greaterEqual(BInteger(0))

    def isNotNatural(self) -> 'BBoolean':
        return self.isNatural()._not()

    def isNatural1(self) -> 'BBoolean':
        return self.greater(BInteger(0))

    def isNotNatural1(self) -> 'BBoolean':
        return self.isNatural1()._not()

    def __hash__(self):
        return hash(31 + self.__value)